### PR TITLE
Server side

### PR DESCRIPTION
**Pull request by noordean**

### CODE REVIEW

File: coverage/lcov-report/index.html [hunk 1]
```diff
@@ -20,24 +20,24 @@
     </h1>
```
Comment (victor-gesit) [inline coverage/lcov-report/index.html]: You're not supposed to upload your coverage folder to git.

File: dist/route/route.js [hunk 2]
```diff
@@ -48,10 +49,17 @@ router.post('/api/create', function (req, res) {
   });
 });
 
+// search endpoint
 router.post('/api/search', search.single(), function (req, res) {
   if (req.body.fileName !== undefined) {
+    if (req.body.searchTerms.length === 0) {
```
Comment (victor-gesit) [inline dist/route/route.js]: Validation shouldn't be done in the route.

File: dist/tests/inverted-index-test.js [hunk 3]
```diff
@@ -1,14 +1,82 @@
...
+var _server2 = _interopRequireDefault(_server);
+
 function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
 
 _invertedIndex2.default.createIndex('book1.json', _invertedIndex2.default.readFile('book1.json'));
 _invertedIndex2.default.createIndex('book2.json', _invertedIndex2.default.readFile('book2.json'));
-_invertedIndex2.default.createIndex('book2.json', _invertedIndex2.default.readFile('invalidBook.json'));
+
+var createdIndexForBook1 = {
+  'book1.json': {
```
Comment (victor-gesit) [inline dist/tests/inverted-index-test.js]: It would be better if you could move all these into an external file and then import the variables.

File: route/route.js [hunk 4]
```diff
@@ -8,31 +8,38 @@ const upload = multer({ dest: 'fixtures/' }).single('fileContent'); // specify t
...
       fs.unlinkSync(path.join('fixtures', req.file.filename)); // delete the uploaded file once the index is created
     }
   });
 });
 
+
+// search endpoint
 router.post('/api/search', search.single(), (req, res) => {
   if (req.body.fileName !== undefined) {
+    if (req.body.searchTerms.length === 0) {
```
Comment (victor-gesit) [inline route/route.js]: Validation should be separated from the route

**Review by ngfizzy — COMMENTED** [review]
88% . Nice one.  You can as well try not to push your coverage file. Generally, auto generated files should not be pushed to github

File: tests/inverted-index-test.js [hunk 5]
```diff
@@ -1,10 +1,180 @@
...
 import indexObj from '../src/inverted-index';
+import app from '../server';
 
 
 indexObj.createIndex('book1.json', indexObj.readFile('book1.json'));
 indexObj.createIndex('book2.json', indexObj.readFile('book2.json'));
-indexObj.createIndex('book2.json', indexObj.readFile('invalidBook.json'));
+
+const createdIndexForBook1 = {
+  'book1.json': {
```
Comment (victor-gesit) [inline tests/inverted-index-test.js]: These files for testing should be put into an external file and imported into the test suite.

**Review by victor-gesit — COMMENTED** [review]
Take a look at the issues I highlighted and fix them, or reply with your explanations.

File: tests/inverted-index-test.js [hunk 6]
```diff
@@ -1,10 +1,180 @@
...
+import path from 'path';
 import indexObj from '../src/inverted-index';
+import app from '../server';
 
 
 indexObj.createIndex('book1.json', indexObj.readFile('book1.json'));
 indexObj.createIndex('book2.json', indexObj.readFile('book2.json'));
-indexObj.createIndex('book2.json', indexObj.readFile('invalidBook.json'));
+
+const createdIndexForBook1 = {
```
Comment (ngfizzy) [inline tests/inverted-index-test.js]: this variable is taking to many lines. you can try to minimize the number of lines it takes by putting your arrays on single line. 

for example. use `for: [0,1]`

File: coverage/lcov-report/index.html (hunk 1 above)
Comment (noordean) [reply coverage/lcov-report/index.html]: Thanks i will correct that

File: dist/route/route.js (hunk 2 above)
Comment (noordean) [reply dist/route/route.js]: Alright i will find a way of validating from my class

File: dist/tests/inverted-index-test.js (hunk 3 above)
Comment (noordean) [reply dist/tests/inverted-index-test.js]: I was thinking of something like that but forgot to implement it. Thanks i will definitely implement it

File: tests/inverted-index-test.js (hunk 5 above)
Comment (noordean) [reply tests/inverted-index-test.js]: Yeah i will

File: coverage/lcov-report/index.html (hunk 1 above)
Comment (nelsonabienorin) [reply coverage/lcov-report/index.html]: Have you corrected that?

File: dist/gulpfile.babel.js [hunk 7]
```diff
@@ -28,6 +28,10 @@ var _gulpCoveralls = require('gulp-coveralls');
```
Comment (nelsonabienorin) [inline dist/gulpfile.babel.js]: Why do you need to bundle your gulp.babel.js here?

File: dist/route/route.js [hunk 8]
```diff
@@ -30,6 +30,7 @@ var upload = (0, _multer2.default)({ dest: 'fixtures/' }).single('fileContent');
 var search = (0, _multer2.default)(); // This is used for search endpoint since no uploading is involved
```
Comment (nelsonabienorin) [inline dist/route/route.js]: Bundling all the files separately will not fully optimize the memory and speed of your app,  bundling into just one file will be advised.

File: src/inverted-index.js [hunk 9]
```diff
@@ -57,8 +55,7 @@ class invertedIndex {
   isValidFileName(fileName) {
     if (fileName.match('.json$') === null) {
```
Comment (nelsonabienorin) [inline src/inverted-index.js]: what does the $ symbol at the end of .json does?

File: src/inverted-index.js [hunk 10]
```diff
@@ -79,8 +76,8 @@ class invertedIndex {
         for (const key in fileContent[i]) {
```
Comment (nelsonabienorin) [inline src/inverted-index.js]: A more descriptive variable name can be used other than key

File: src/inverted-index.js [hunk 11]
```diff
@@ -79,8 +76,8 @@ class invertedIndex {
...
           const getTokensPerKey = [];
           const getEachKey = fileContent[i][key].toLowerCase().replace(/\W+/g, ' ').split(' ');
```
Comment (nelsonabienorin) [inline src/inverted-index.js]: A more descriptive variable name can be used

File: src/inverted-index.js [hunk 12]
```diff
@@ -96,12 +93,11 @@ class invertedIndex {
       getAllUniqueTokens.forEach((term) => {
         if (innerIndex[term] !== undefined) {
           innerIndex[term] = [...innerIndex[term], Number(innerIndex[term][innerIndex[term].length - 1] + 1)];
-          }
-        else {
-          for (let l = 0; l < fileContent.length; l += 1){
-            for (const keyy in fileContent[l]){
+        } else {
+          for (let l = 0; l < fileContent.length; l += 1) {
+            for (const keyy in fileContent[l]) {
```
Comment (nelsonabienorin) [inline src/inverted-index.js]: keyy ? Why is this used as a variable name and also why do you use const?

File: src/inverted-index.js [hunk 13]
```diff
@@ -131,29 +126,25 @@ class invertedIndex {
 
     if (Object.keys(index).length === 0) {
       return { error: 'Index has not been created. Kindly create index before searching' };
-    }
-    else {
+    } else {
       if (Array.isArray(fileName)) {
         getSearchTerms.push(...fileName); // definitly the second arg is seachTerms
```
Comment (nelsonabienorin) [inline src/inverted-index.js]: change definitly to definitely

File: tests/inverted-index-test.js [hunk 14]
```diff
@@ -1,10 +1,180 @@
-
+import supertest from 'supertest';
+import path from 'path';
 import indexObj from '../src/inverted-index';
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: whats indexObj? Can the name be more descriptive?

File: tests/inverted-index-test.js [hunk 15]
```diff
@@ -42,7 +212,7 @@ describe('Search index', () => {
   });
 
   it('should return { "understand":[0,1], "from": [1] } for array ["understand","from"]', () => {
-    const result = { 'book2.json': { 'from': [1], 'understand': [0, 1] }, 'book1.json': { 'from': [-1], 'understand': [-1] } };
+    const result = { 'book2.json': { from: [1], understand: [0, 1] }, 'book1.json': { from: [-1], understand: [-1] } };
     expect(indexObj.searchIndex(indexObj.index, ['understand', 'from'])).toEqual(result);
   });
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: Apart from testing for a json array, you should also test for an empty json array

File: tests/inverted-index-test.js [hunk 16]
```diff
@@ -51,6 +221,105 @@ describe('Search index', () => {
...
 
   it('should return [-1] for any word not present', () => {
-    expect(indexObj.searchIndex(indexObj.index, 'book1.json', 'catwalk')).toEqual({ 'catwalk': [-1] });
+    expect(indexObj.searchIndex(indexObj.index, 'book1.json', 'catwalk')).toEqual({ catwalk: [-1] });
+  });
+});
+
+
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: Where is the test for an Invalid file?

File: tests/inverted-index-test.js [hunk 17]
```diff
@@ -51,6 +221,105 @@ describe('Search index', () => {
...
+    .expect({ error: 'Index has not been created for the specified file' })
+        .end((err, res) => {
+          if (err) {
+            done.fail(err);
+          } else {
+            done();
+          }
+        });
+  });
+
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: I can't find the test for "Ensure passed in index is in the valid format"

File: tests/inverted-index-test.js [hunk 18]
```diff
@@ -51,6 +221,105 @@ describe('Search index', () => {
...
+  it('should return correct result if both fileName and searchTerms are supplied', (done) => {
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: Can't find the test for "Ensures index returns the correct results when searched"

File: tests/inverted-index-test.js [hunk 19]
```diff
@@ -51,6 +221,105 @@ describe('Search index', () => {
...
+    .post('/api/search')
+    .send({ fileName: 'book1.json', searchTerms: 'best' })
+    .expect({ best: [0] })
+        .end((err, res) => {
+          if (err) {
+            done.fail(err);
+          } else {
+            done();
+          }
+        });
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js:188]: Can't find test that "Ensure searchIndex can handle an array of search terms"

File: tests/inverted-index-test.js [hunk 20]
```diff
@@ -51,6 +221,105 @@ describe('Search index', () => {
...
+  });
+
```
Comment (nelsonabienorin) [inline tests/inverted-index-test.js]: Can't find test that "Ensure searchIndex can handle a varied number of search terms as arguments"

File: tests/super-test.js [hunk 21]
```diff
@@ -1,38 +0,0 @@
-import supertest from 'supertest';
```
Comment (nelsonabienorin) [inline tests/super-test.js:1]: What's the essence of this file in your test?

**Review by nelsonabienorin — APPROVED** [review]
Good work done so far but kindly apply all the requested changes and reply where necessary.

File: dist/gulpfile.babel.js (hunk 7 above)
Comment (noordean) [reply dist/gulpfile.babel.js]: It was actually a mistake. It's been corrected

File: dist/route/route.js (hunk 8 above)
Comment (noordean) [reply dist/route/route.js]: It's been corrected

File: tests/inverted-index-test.js (hunk 6 above)
Comment (noordean) [reply tests/inverted-index-test.js]: Thanks. Implemented

File: src/inverted-index.js (hunk 9 above)
Comment (noordean) [reply src/inverted-index.js]: The $ helps check for a value at the end of a certain string, so it enables me to check for '.json' at the  end of a file name

File: src/inverted-index.js (hunk 10 above)
Comment (noordean) [reply src/inverted-index.js]: I had already refactored this

File: src/inverted-index.js (hunk 11 above)
Comment (noordean) [reply src/inverted-index.js]: Done

File: src/inverted-index.js (hunk 12 above)
Comment (noordean) [reply src/inverted-index.js]: Already refactored

File: src/inverted-index.js (hunk 13 above)
Comment (noordean) [reply src/inverted-index.js]: Done

File: tests/inverted-index-test.js (hunk 14 above)
Comment (noordean) [reply tests/inverted-index-test.js]: It's been changed to InvertedIndexObject

File: tests/inverted-index-test.js (hunk 15 above)
Comment (noordean) [reply tests/inverted-index-test.js]: This has been implemented

File: tests/inverted-index-test.js (hunk 19 above)
Comment (noordean) [reply tests/inverted-index-test.js:188]: It's been included

File: tests/super-test.js (hunk 21 above)
Comment (noordean) [reply tests/super-test.js:1]: It has been removed. It's no more needed

File: dist/route/route.js (hunk 8 above)
Comment (noordean) [reply dist/route/route.js]: You mean i actually transpile all the files into a single file??

File: README.md [hunk 22]
```diff
@@ -27,4 +28,6 @@ The API can be tested with postman as described below:
   - fileContent: It is the file itself
 - Search for words by making another request to localhost:3000/api/search with the following form-data bodies:
   - fileName (Optional): The document you want to search from
-  - searchTerms: The words you are searching. It can be an array of words.
\ No newline at end of file
+  - searchTerms: The words you are searching. It can be an array of words.
+  
+Note: [-1] denotes 'word not found'
```
Comment (Seyi-Keye) [inline README.md]: Well done so far with the content of your README file, however, to better the user of your application better understand how to navigate and use this application, you might want to go a bit into details on how to install, clone, contribute, and dependencies used admist others.

Also, you might want to also put an example file of what a valid file content should look like and examples of the result to expect from the two API endpoints when hit.

File: dist/route/endpoints-validation.js [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+          return { error: 'The searchTerms cannot be empty' };
+        }
+        return _invertedIndex2.default.searchIndex(index, searchTerms);
+      }
+    }
+  }]);
+
+  return Validation;
+}();
+
```
Comment (Seyi-Keye) [inline dist/route/endpoints-validation.js]: Like i earlier mentioned, the bundled folder can be ignored

File: src/inverted-index.js [hunk 24]
```diff
@@ -21,32 +23,35 @@ class invertedIndex {
...
+    this.file = fileName;
     try {
       JSON.parse(fs.readFileSync(path.join('fixtures', fileName)));
+    } catch (e) {
+      return 'Invalid JSON file';
     }
-    catch (e) {
+    if (JSON.parse(fs.readFileSync(path.join('fixtures', fileName))).length === 0) {
+      return 'Empty JSON file';
+    } else if (JSON.parse(fs.readFileSync(path.join('fixtures', fileName)))[0].title === undefined ||
```
Comment (Seyi-Keye) [inline src/inverted-index.js]: For better code readablity, you can use and obey style guides

File: src/inverted-index.js [hunk 25]
```diff
@@ -129,76 +151,65 @@ class invertedIndex {
...
       if (Array.isArray(fileName)) {
         getSearchTerms.push(...fileName); // definitly the second arg is seachTerms
-      }
-      else {
-        getSearchTerms.push(fileName); // definitly the second arg is fileName
+      } else {
+        getSearchTerms.push(fileName); // definitely the second arg is fileName
       }
 
       for (let i = 0; i < searchTerms.length; i += 1) {
```
Comment (Seyi-Keye) [inline src/inverted-index.js:200]: A forEach is preferred to a for loop when following best practices.

File: tests/inverted-index-test.js [hunk 26]
```diff
@@ -1,56 +1,262 @@
...
+  it('should return true for a valid index', () => {
+    expect(invertedIndexObject.isIndexValid(invertedIndexObject.createIndex('book1.json', invertedIndexObject.readFile('book1.json')))).toBe(true);
+  });
+
+  it('should return false for an invalid index', () => {
+    expect(invertedIndexObject.isIndexValid(invertedIndexObject.createIndex('invalidJsonBook.json', invertedIndexObject.readFile('invalidJsonBook.json')))).toBe(false);
+  });
+});
+
+describe('create index endpoint', () => {
```
Comment (Seyi-Keye) [inline tests/inverted-index-test.js]: We always advice seperating concerns and you have a done a good jod with your test, however, you can better seperate instances by placing API test and Class test in different files

**Review by Seyi-Keye — CHANGES_REQUESTED** [review]
